jps-plugin/src/org/intellij/erlang/jps/builder/ErlangModuleBuildOrderBuilder.java

```diff
@@ -0,0 +1,199 @@
+    return getSortedDirtyModules(projectBuildOrder.myErlangFiles, allDirtyFiles);
+  }
+
+
+  private static void addPrepareDependenciesFailedMessage(@NotNull CompileContext context) {
+    context.processMessage(new CompilerMessage(NAME, BuildMessage.Kind.WARNING, "The project will be fully rebuilt due to errors."));
+  }
+
+  private static Set<String> getAllDirtyFiles(@NotNull List<String> dirtyFiles,
+                                              @NotNull GraphGenerator<String> graph) {
+    Set<String> allDirtyFiles = ContainerUtil.newHashSet();
+    for (String node : dirtyFiles) {
+      if (node != null) {
+        findDirtyFiles(node, graph, allDirtyFiles);
+      }
+    }
+    return allDirtyFiles;
+  }
+
+  private static void findDirtyFiles(@NotNull String filePath,
+                                     @NotNull GraphGenerator<String> graph,
+                                     @NotNull Set<String> dirtyFiles) {
+    if (dirtyFiles.contains(filePath)) return;
+    dirtyFiles.add(filePath);
+    Iterator<String> childIterator = graph.getOut(filePath);
+    while (childIterator.hasNext()) {
+      findDirtyFiles(childIterator.next(), graph, dirtyFiles);
+    }
+  }
+
+  @NotNull
+  private static List<String> getSortedDirtyModules(@NotNull List<ErlangFileDescriptor> sortedFiles,
+                                                    @NotNull final Set<String> allDirtyFiles) {
+    return ContainerUtil.mapNotNull(sortedFiles, new Function<ErlangFileDescriptor, String>() {
+      @Nullable
+      @Override
+      public String fun(ErlangFileDescriptor node) {
+        return node.myPath.endsWith(".erl") && allDirtyFiles.contains(node.myPath) ? node.myPath : null;
+      }
+    });
+  }
+
+  private static void addFilesToBuiltTarget(@NotNull CompileContext context,
+                                            @NotNull List<String> sortedDirtyErlangModules) {
+    for (String filePath : sortedDirtyErlangModules) {
+      ErlangSourceRootDescriptor root = getSourceRootDescriptor(context, filePath);
+      if (root == null) {
+        LOG.error("Source root not found.");
+        return;
+      }
+      ErlangTarget target = (ErlangTarget) root.getTarget();
+
+      if (root.isTests()) {
+        target.getBuildOrder().myOrderedErlangTestFilePaths.add(filePath);
+      }
+      else {
+        target.getBuildOrder().myOrderedErlangFilePaths.add(filePath);
+      }
+    }
+  }
+
+  @Nullable
+  private static ErlangSourceRootDescriptor getSourceRootDescriptor(@NotNull CompileContext context, String filePath) {
+    List<ErlangTargetType> targetTypes = Collections.singletonList(ErlangTargetType.INSTANCE);
+    return context.getProjectDescriptor().getBuildRootIndex().findParentDescriptor(new File(filePath),
+                                                                                   targetTypes,
+                                                                                   context);
```

Comment:
It would be better to inline this method, because you'd extract a variable for build root index, and not create new instances of this singleton list every time you do an index lookup.